core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -211,7 +211,23 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
       List<Expression> expressions) {
     final List<Expression> list = new ArrayList<>();
     for (int i = 0; i < expressions.size(); i++) {
-      list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      if (targetTypes[i].isArray()) {
+        // only the last type could be variable length argument. All left expressions
+        // are supposed to be the same type as targetTypes[i].getComponentType().
+        List<Expression> subList = expressions.subList(i, expressions.size());
+        list.addAll(fromInternal(targetTypes[i].getComponentType(), subList));
+        break;
+      } else {
+        list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      }
+    }
+    return list;
+  }
+
+  static List<Expression> fromInternal(Class<?> targetType, List<Expression> expressions) {
```

Comment:
Does it need to be package private? Can't we make it just private?

Reply:
Updated